File: leetcode/jump_game.py
from typing import List 


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        goal = len(nums) - 1
        for i in range(len(nums))[::-1]:
            if i + nums[i] >= goal:
                goal = i
        return not goal
    
if __name__ == "__main__":
    s = Solution()
    nums = [2, 3, 1, 1, 4]
    # nums = [3, 2, 1, 0, 4]
    print(s.canJump(nums))
            

File: leetcode/jump_game.cpp
using namespace std;
#include <iostream>
#include <vector>


class Solution {
public: 
    bool canJump(vector<int>& nums) {
        int goal = nums.size() - 1;
        for (int i = goal; i >= 0; i--) {
            if (i + nums[i] >= goal) {
                goal = i;
            }
        }
        return goal == 0;
    }
};

int main() {
    Solution s;
    vector<int> nums = {2, 3, 1, 1, 4};
    cout << s.canJump(nums) << endl;
}